File: android/app/src/main/kotlin/com/bockmedia/console/local/ClientPrefsSync.kt
package com.bockmedia.console.local

import android.content.Context
import com.bockmedia.console.BockMediaApp
import com.bockmedia.console.data.api.bockJson
import com.bockmedia.console.data.api.dto.ClientPrefsResponse
import com.bockmedia.console.data.local.AppPreferences
import com.bockmedia.console.data.repository.BockMediaRepository
import com.bockmedia.console.domain.model.HomeFeedCache
import com.bockmedia.console.domain.model.HomeLoadCoordinator
import com.bockmedia.console.domain.model.RESUME_PULL_DEBOUNCE_MS
import com.bockmedia.console.domain.model.shouldRefreshHomeForProfile
import com.bockmedia.console.domain.model.shouldSkipResumePull
import com.bockmedia.console.domain.model.HomeTileEngagement
import com.bockmedia.console.local.HomeSectionPinsStore
import com.bockmedia.console.domain.model.LibrarySort
import com.bockmedia.console.domain.model.LibrarySessionCache
import com.bockmedia.console.domain.model.LibraryViewMode
import com.bockmedia.console.domain.model.SearchBrowseSessionCache
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/** Syncs phone settings and preferences to the server (per household profile). */
object ClientPrefsSync {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var pushJob: Job? = null
    @Volatile private var pulling = false
    @Volatile private var lastPullCompletedMs: Long = 0L

    private val _profileChangeRevision = MutableStateFlow(0)
    val profileChangeRevision: StateFlow<Int> = _profileChangeRevision.asStateFlow()

    private fun bumpProfileRevision() {
        _profileChangeRevision.value += 1
    }

    fun schedulePush(context: Context) {
        pushJob?.cancel()
        pushJob = scope.launch {
            delay(900)
            runCatching { push(context.applicationContext) }
        }
    }

    fun markBootPullCompleted() {
        lastPullCompletedMs = System.currentTimeMillis()
    }

    fun shouldSkipResumePull(): Boolean =
        shouldSkipResumePull(lastPullCompletedMs, System.currentTimeMillis(), RESUME_PULL_DEBOUNCE_MS)

    suspend fun pullAndApply(context: Context, profileSwitch: Boolean = false) {
        if (pulling) return
        pulling = true
        try {
            val app = BockMediaApp.get(context)
            val repository = app.repository
            val prefs = app.preferences
            val clientId = ClientIdStore.clientId(context)
            val profileAdjusted = runCatching {
                syncHouseholdProfile(context, repository, clientId)
            }.getOrElse { return }
            var memberId = ActiveProfileStore.activeMemberId(context)
            val remote = runCatching { repository.clientPrefs(clientId, memberId) }
                .getOrElse { return }
            val merged = if (profileSwitch) remote.memberPrefs else remote.merged
            if (profileAdjusted && memberId.isNullOrBlank() &&
                !ActiveProfileStore.hasProfileChoice(context)
            ) {
                merged["activeMemberId"]?.jsonPrimitive?.content?.trim()?.takeIf { it.isNotBlank() }
                    ?.let { fromPrefs ->
                        if (HouseholdStore.memberExists(fromPrefs)) {
                            ActiveProfileStore.setActiveMember(context, fromPrefs)
                            memberId = fromPrefs
                        }
                    }
            }
            applyMerged(context, prefs, remote, profileSwitch = profileSwitch)
            memberId = ActiveProfileStore.activeMemberId(context)
            if (!memberId.isNullOrBlank()) {
                runCatching {
                    repository.bindClient(
                        clientId,
                        memberId,
                        InstallIdentity.phoneId(context),
                    )
                }
            }
            repository.clearRatingsCache()
            if (profileSwitch || shouldRefreshHomeForProfile(
                    !ActiveProfileStore.activeMemberId(context).isNullOrBlank(),
                    HomeFeedCache.peek(),
                    HomeFeedCache.peekHasRatedSongs(),
                )
            ) {
                HomeFeedCache.invalidate()
                HomeLoadCoordinator.resetReloadWindow()
            }
            lastPullCompletedMs = System.currentTimeMillis()
        } finally {
            pulling = false
        }
    }

    private suspend fun syncHouseholdProfile(
        context: Context,
        repository: BockMediaRepository,
        clientId: String,
    ): Boolean {
        val household = HouseholdStore.refresh(repository)
        val beforeMember = ActiveProfileStore.activeMemberId(context)
        rebindFromPhone(context, repository, clientId, household)
        var changed = reconcileActiveMember(context, household)
        if (restoreActiveMember(context, household, clientId)) changed = true
        if (beforeMember != ActiveProfileStore.activeMemberId(context)) changed = true
        if (changed) bumpProfileRevision()
        return changed
    }

    private fun reconcileActiveMember(context: Context, household: com.bockmedia.console.data.api.dto.HouseholdResponse): Boolean {
        val mid = ActiveProfileStore.activeMemberId(context)?.trim().orEmpty()
        if (mid.isEmpty() || household.members.any { it.id == mid }) return false
        ActiveProfileStore.clearStaleMember(context)
        return true
    }

    private fun restoreActiveMember(
        context: Context,
        household: com.bockmedia.console.data.api.dto.HouseholdResponse,
        clientId: String,
    ): Boolean {
        if (!ActiveProfileStore.activeMemberId(context).isNullOrBlank()) return false
        val deviceId = clientDeviceId(clientId)
        val fromBinding = household.clientBindings.firstOrNull { it.clientDeviceId == deviceId }?.memberId
        if (!fromBinding.isNullOrBlank() && household.members.any { it.id == fromBinding }) {
            ActiveProfileStore.setActiveMember(context, fromBinding)
            return true
        }
        if (!ActiveProfileStore.hasProfileChoice(context) && household.members.size == 1) {
            household.members.firstOrNull()?.id?.takeIf { it.isNotBlank() }?.let {
                ActiveProfileStore.setActiveMember(context, it)
                return true
            }
        }
        return false
    }

    private fun clientDeviceId(clientId: String): String {
        val cid = clientId.trim()
        return if (cid.isEmpty()) "" else "client-$cid"
    }

    private suspend fun rebindFromPhone(
        context: Context,
        repository: BockMediaRepository,
        clientId: String,
        household: com.bockmedia.console.data.api.dto.HouseholdResponse,
    ) {
        if (!ActiveProfileStore.activeMemberId(context).isNullOrBlank()) return
        val phoneId = InstallIdentity.phoneId(context)
        if (phoneId.isBlank()) return
        val model = android.os.Build.MODEL?.trim().orEmpty()
        val label = if (model.isNotBlank()) "Android · $model" else "This phone"
        val memberId = runCatching {
            repository.connectInstall(phoneId, label, clientId)
        }.getOrNull()?.trim().orEmpty()
        if (memberId.isNotBlank() && household.members.any { it.id == memberId }) {
            ActiveProfileStore.setActiveMember(context, memberId)
        }
    }

    /** No-op — profile is chosen explicitly in [ProfilePickerGate] or Family. */
    suspend fun ensureProfileLinked(context: Context): Boolean = false

    suspend fun push(context: Context, memberIdOverride: String? = null) {
        val app = BockMediaApp.get(context)
        val repository = app.repository
        val prefs = app.preferences
        val clientId = ClientIdStore.clientId(context)
        val memberId = memberIdOverride?.takeIf { it.isNotBlank() }
            ?: ActiveProfileStore.activeMemberId(context)
        repository.putClientPrefs(
            clientId = clientId,
            memberId = memberId,
            memberPrefs = collectMemberPrefs(context, prefs, memberId),
            clientPrefs = collectClientPrefs(context),
        )
    }

    suspend fun onActiveMemberChanged(
        context: Context,
        memberId: String?,
        previousMemberId: String?,
    ) {
        val app = BockMediaApp.get(context)
        val clientId = ClientIdStore.clientId(context)
        // Persist locally first so the picker dismisses even if sync is slow.
        if (memberId.isNullOrBlank()) {
            ActiveProfileStore.chooseUnattributed(context)
        } else {
            ActiveProfileStore.setActiveMember(context, memberId)
        }
        OfflineDownloadManager.onActiveProfileChanged(context, previousMemberId)
        app.repository.clearRatingsCache()
        app.repository.invalidatePlaylistsCache()
        HomeFeedCache.invalidate()
        HomeLoadCoordinator.resetReloadWindow()
        LibrarySessionCache.invalidate()
        SearchBrowseSessionCache.invalidate()
        bumpProfileRevision()
        // Save outgoing profile to the server before binding the new one.
        if (!previousMemberId.isNullOrBlank()) {
            runCatching { push(context, previousMemberId) }
        }
        runCatching { app.repository.bindClient(clientId, memberId, InstallIdentity.phoneId(context)) }
        runCatching { pullAndApply(context, profileSwitch = true) }
    }

    private suspend fun collectMemberPrefs(
        context: Context,
        prefs: AppPreferences,
        memberId: String?,
    ): JsonObject = buildJsonObject {
        put("searchAllLibraries", prefs.isSearchAllLibrariesSync())
        prefs.getSearchSourcePathSync()?.let { put("searchSourcePath", it) }
        put("downloadWifiOnly", prefs.isDownloadWifiOnlySync())
        put("crossfadeSeconds", prefs.getCrossfadeSecondsSync())
        put("continueAfterQueue", prefs.getContinueAfterQueueSync())
        put("nowPlayingVideo", prefs.isNowPlayingVideoSync())
        put("rememberMe", prefs.isRememberMeSync())
        memberId?.let { put("activeMemberId", it) }
        put("searchSelections", encodeSearchSelections(SearchHistoryStore(context).selectionsSync()))
        HomeTileEngagement.exportJson()?.let { put("homeTileEngagement", JsonPrimitive(it)) }
        HomeSectionPinsStore.exportJson()?.let { put("homeSectionPins", JsonPrimitive(it)) }
        if (!memberId.isNullOrBlank()) {
            LastDeviceStore(context).lastDeviceSync()?.let { put("lastDevice", it) }
            val pinned = PinnedDevicesStore(context).pinnedValuesSync()
            if (pinned.isNotEmpty()) {
                put("pinnedDevices", buildJsonArray { pinned.forEach { add(JsonPrimitive(it)) } })
            }
            val offline = OfflineDownloadSync.recordsForMember(context, memberId)
            if (offline.isNotEmpty()) {
                put("offlineDownloads", bockJson.parseToJsonElement(OfflineDownloadSync.encode(offline)))
            }
            val library = LibraryPrefsStore(context).loadSync()
            put("libraryTab", LibraryPrefsStore.filterTabValue(library.filter))
            put("libraryViewMode", if (library.viewMode == LibraryViewMode.Grid) "grid" else "list")
            put("librarySortBy", if (library.sort == LibrarySort.Recents) "recents" else "name")
            put("librarySortOrder", if (library.sort == com.bockmedia.console.domain.model.LibrarySort.Recents) "desc" else "asc")
        }
    }

    private suspend fun collectClientPrefs(context: Context): JsonObject = buildJsonObject {
        // Device playback prefs live on the household member (survive reinstall).
    }

    private suspend fun applyMerged(
        context: Context,
        prefs: AppPreferences,
        remote: ClientPrefsResponse,
        profileSwitch: Boolean = false,
    ) {
        val merged = if (profileSwitch) remote.memberPrefs else remote.merged
        if (merged.isEmpty() && !profileSwitch) return
        merged["searchAllLibraries"]?.jsonPrimitive?.booleanOrNull?.let {
            prefs.setSearchAllLibraries(it)
        } ?: run {
            if (profileSwitch) prefs.setSearchAllLibraries(true)
        }
        when {
            merged["searchSourcePath"]?.jsonPrimitive?.content?.isNotBlank() == true -> {
                prefs.setSearchSourcePath(merged["searchSourcePath"]!!.jsonPrimitive.content)
            }
            profileSwitch || merged.containsKey("searchSourcePath") -> {
                prefs.setSearchSourcePath(null)
            }
        }
        merged["downloadWifiOnly"]?.jsonPrimitive?.booleanOrNull?.let {
            prefs.setDownloadWifiOnly(it)
        } ?: run {
            if (profileSwitch) prefs.setDownloadWifiOnly(false)
        }
        merged["crossfadeSeconds"]?.jsonPrimitive?.intOrNull?.let {
            prefs.setCrossfadeSeconds(it)
        } ?: run {
            if (profileSwitch) prefs.setCrossfadeSeconds(0)
        }
        merged["continueAfterQueue"]?.jsonPrimitive?.content?.takeIf { it.isNotBlank() }?.let {
            prefs.setContinueAfterQueue(it)
        } ?: run {
            if (profileSwitch) prefs.setContinueAfterQueue("off")
        }
        merged["nowPlayingVideo"]?.jsonPrimitive?.booleanOrNull?.let {
            prefs.setNowPlayingVideo(it)
        } ?: run {
            if (profileSwitch) prefs.setNowPlayingVideo(false)
        }
        merged["rememberMe"]?.jsonPrimitive?.booleanOrNull?.let {
            prefs.setRememberMe(it)
        }
        decodeSearchSelections(merged["searchSelections"])?.let { items ->
            SearchHistoryStore(context).replaceSelections(items)
        }
        merged["homeTileEngagement"]?.jsonPrimitive?.content?.let {
            HomeTileEngagement.importJson(it)
        }
        merged["homeSectionPins"]?.jsonPrimitive?.content?.let {
            HomeSectionPinsStore.importJson(it)
        }
        merged["lastDevice"]?.jsonPrimitive?.content?.takeIf { it.isNotBlank() }?.let {
            LastDeviceStore(context).setLastDevice(it)
        }
        decodeStringList(merged["pinnedDevices"])?.let { pinned ->
            PinnedDevicesStore(context).setPinned(pinned)
        }
        val offlineRecords = decodeOfflineDownloads(merged["offlineDownloads"])
        if (offlineRecords != null) {
            OfflineDownloadSync.applyRemote(context, offlineRecords)
        } else if (profileSwitch) {
            OfflineDownloadSync.applyRemote(context, emptyList())
        }
        if (merged.containsKey("libraryTab") ||
            merged.containsKey("libraryViewMode") ||
            merged.containsKey("librarySortBy")
        ) {
            LibraryPrefsStore(context).applyRemote(
                tab = merged["libraryTab"]?.jsonPrimitive?.content,
                viewMode = merged["libraryViewMode"]?.jsonPrimitive?.content,
                sortBy = merged["librarySortBy"]?.jsonPrimitive?.content,
                sortOrder = merged["librarySortOrder"]?.jsonPrimitive?.content,
            )
        }
    }

    private fun encodeSearchSelections(items: List<SearchRecentSelection>): JsonElement {
        val serializer = ListSerializer(SearchRecentSelection.serializer())
        return bockJson.parseToJsonElement(bockJson.encodeToString(serializer, items))
    }

    private fun decodeSearchSelections(element: JsonElement?): List<SearchRecentSelection>? {
        if (element == null || element !is JsonArray) return null
        return runCatching {
            bockJson.decodeFromString(
                ListSerializer(SearchRecentSelection.serializer()),
                element.toString(),
            )
        }.getOrNull()
    }

    private fun decodeOfflineDownloads(element: JsonElement?): List<OfflineDownloadRecord>? {
        if (element == null || element !is JsonArray) return null
        return OfflineDownloadSync.decode(element.toString())
    }

    private fun decodeStringList(element: JsonElement?): List<String>? {
        if (element == null || element !is JsonArray) return null
        return element.mapNotNull { it.jsonPrimitive.content.takeIf { s -> s.isNotBlank() } }
    }
}

File: ios/BockMedia/Data/Local/ClientPrefsSync.swift
import Foundation
import UIKit

enum ClientPrefsSyncNotifications {
    static let profileChanged = Notification.Name("BockProfileChanged")
    static let prefsApplied = Notification.Name("BockPrefsApplied")
}

/// Syncs phone settings and preferences to the server (per household profile).
@MainActor
enum ClientPrefsSync {
    private static weak var repository: BockMediaRepository?
    private static var pushTask: Task<Void, Never>?
    private static var pulling = false
    private static var lastPullCompletedMs: Int64 = 0
    private(set) static var profileChangeRevision = 0

    static func configure(repository: BockMediaRepository) {
        self.repository = repository
    }

    static func schedulePush() {
        guard let repository else { return }
        schedulePush(repository: repository)
    }

    static func schedulePush(repository: BockMediaRepository) {
        pushTask?.cancel()
        pushTask = Task {
            try? await Task.sleep(nanoseconds: 900_000_000)
            try? await push(repository: repository)
        }
    }

    static func markBootPullCompleted() {
        lastPullCompletedMs = Int64(Date().timeIntervalSince1970 * 1000)
    }

    static func shouldSkipResumePull() -> Bool {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        return HomeLoadCoordinator.shouldSkipResumePull(
            lastPullCompletedMs: lastPullCompletedMs,
            nowMs: now
        )
    }

    static func pullAndApply(repository: BockMediaRepository, profileSwitch: Bool = false) async {
        if pulling {
            if profileSwitch {
                while pulling {
                    try? await Task.sleep(nanoseconds: 100_000_000)
                }
            } else {
                _ = await HouseholdStore.refresh(repository: repository, force: true)
                AutomationSessionCache.invalidate()
                return
            }
        }
        pulling = true
        defer { pulling = false }
        let prefs = repository.preferences
        let clientId = ClientIdStore.clientId()
        let profileAdjusted = await syncHouseholdProfile(repository: repository, clientId: clientId, profileSwitch: profileSwitch)
        let memberId = ActiveProfileStore.activeMemberId()
        if let remote = try? await repository.clientPrefs(clientId: clientId, memberId: memberId) {
            // When a profile is active, read profile-scoped prefs from the member bucket (not stale client merge).
            let merged = (profileSwitch || memberId != nil) ? remote.memberPrefsDict : remote.mergedDict
            if profileAdjusted, ActiveProfileStore.activeMemberId() == nil,
               !ActiveProfileStore.hasProfileChoice(),
               let fromPrefs = merged["activeMemberId"] as? String,
               HouseholdStore.memberExists(fromPrefs) {
                ActiveProfileStore.setActiveMember(fromPrefs)
            }
            applyMerged(prefs: prefs, merged: merged, repository: repository, profileSwitch: profileSwitch)
        }
        if let memberId = ActiveProfileStore.activeMemberId(), !memberId.isEmpty {
            try? await repository.bindClient(clientId: clientId, memberId: memberId, phoneId: InstallIdentity.phoneId())
        }
        repository.clearRatingsCache()
        if profileSwitch || shouldRefreshHomeForProfile(
            activeProfileLinked: ActiveProfileStore.activeMemberId() != nil,
            feed: HomeFeedCache.peek(),
            hasRatedSongs: HomeFeedCache.peekHasRatedSongs()
        ) {
            HomeFeedCache.invalidate()
            HomeLoadCoordinator.resetReloadWindow()
        }
        lastPullCompletedMs = Int64(Date().timeIntervalSince1970 * 1000)
    }

    /// No-op — profile is chosen explicitly in ProfilePickerGate or Family.
    static func ensureProfileLinked(repository: BockMediaRepository) async -> Bool { false }

    static func onActiveMemberChanged(
        repository: BockMediaRepository,
        memberId: String?,
        previousMemberId: String? = nil
    ) async {
        pushTask?.cancel()
        pushTask = nil
        let clientId = ClientIdStore.clientId()
        if let memberId, !memberId.isEmpty {
            ActiveProfileStore.setActiveMember(memberId)
        } else {
            ActiveProfileStore.chooseUnattributed()
        }
        if let previousMemberId, !previousMemberId.isEmpty {
            try? await push(repository: repository, memberIdOverride: previousMemberId)
        }
        try? await repository.bindClient(clientId: clientId, memberId: memberId, phoneId: InstallIdentity.phoneId())
        await pullAndApply(repository: repository, profileSwitch: true)

        OfflineDownloadManager.shared.onActiveProfileChanged(previousMemberId: previousMemberId)
        OfflineDownloadSync.claimOrphansForActiveProfile()
        repository.clearRatingsCache()
        HomeFeedCache.invalidate()
        HomeLoadCoordinator.resetReloadWindow()
        LibrarySessionCache.invalidate()
        LibraryCachePersistence.clear()
        SessionDataStore.invalidatePlaylists()
        SearchBrowseSessionCache.invalidate()
        SearchQueryCache.invalidate()
        SearchResultsSessionCache.clear()
        AutomationSessionCache.invalidate()
        profileChangeRevision += 1
        NotificationCenter.default.post(name: ClientPrefsSyncNotifications.profileChanged, object: nil)
    }

    static func push(repository: BockMediaRepository, memberIdOverride: String? = nil) async throws {
        let prefs = repository.preferences
        let clientId = ClientIdStore.clientId()
        let memberId = memberIdOverride?.nilIfBlank ?? ActiveProfileStore.activeMemberId()
        _ = try await repository.putClientPrefs(
            clientId: clientId,
            memberId: memberId,
            memberPrefs: collectMemberPrefs(prefs: prefs, memberId: memberId),
            clientPrefs: [:]
        )
    }

    // MARK: - Private

    /// Pull household from server, reconcile stale profiles, restore phone/binding picks.
    @discardableResult
    private static func syncHouseholdProfile(
        repository: BockMediaRepository,
        clientId: String,
        profileSwitch: Bool
    ) async -> Bool {
        let household = await HouseholdStore.refresh(
            repository: repository,
            force: profileSwitch || HouseholdStore.members().isEmpty
        )
        let beforeMember = ActiveProfileStore.activeMemberId()
        await rebindFromPhone(repository: repository, clientId: clientId, household: household)
        var changed = reconcileActiveMember(household)
        if restoreActiveMember(household, clientId: clientId) {
            changed = true
        }
        if ActiveProfileStore.activeMemberId() != beforeMember {
            changed = true
        }
        if changed {
            profileChangeRevision += 1
            NotificationCenter.default.post(name: ClientPrefsSyncNotifications.profileChanged, object: nil)
        }
        return changed
    }

    private static func reconcileActiveMember(_ household: HouseholdResponse) -> Bool {
        guard let mid = ActiveProfileStore.activeMemberId(), !HouseholdStore.memberExists(mid) else { return false }
        ActiveProfileStore.clearStaleMember()
        return true
    }

    private static func restoreActiveMember(_ household: HouseholdResponse, clientId: String) -> Bool {
        guard ActiveProfileStore.activeMemberId() == nil else { return false }
        let deviceId = clientDeviceId(clientId)
        if let fromBinding = household.clientBindings.first(where: { $0.clientDeviceId == deviceId })?.memberId,
           HouseholdStore.memberExists(fromBinding) {
            ActiveProfileStore.setActiveMember(fromBinding)
            return true
        }
        if !ActiveProfileStore.hasProfileChoice(),
           household.members.count == 1,
           let only = household.members.first?.id {
            ActiveProfileStore.setActiveMember(only)
            return true
        }
        return false
    }

    private static func clientDeviceId(_ clientId: String) -> String {
        let cid = clientId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cid.isEmpty else { return "" }
        return "client-\(cid)"
    }

    private static func rebindFromPhone(
        repository: BockMediaRepository,
        clientId: String,
        household: HouseholdResponse
    ) async {
        if ActiveProfileStore.activeMemberId() != nil { return }
        let phoneId = InstallIdentity.phoneId()
        guard !phoneId.isEmpty else { return }
        let model = UIDevice.current.model.trimmingCharacters(in: .whitespacesAndNewlines)
        let label = model.isEmpty ? "This phone" : "iOS · \(model)"
        guard let memberId = try? await repository.connectInstall(
            phoneId: phoneId,
            deviceName: label,
            clientId: clientId
        ), household.members.contains(where: { $0.id == memberId }) else { return }
        ActiveProfileStore.setActiveMember(memberId)
    }

    private static func collectMemberPrefs(prefs: AppPreferences, memberId: String?) -> [String: Any] {
        var out: [String: Any] = [
            "searchAllLibraries": prefs.searchAllLibraries,
            "downloadWifiOnly": prefs.downloadWifiOnly,
            "crossfadeSeconds": prefs.crossfadeSeconds,
            "continueAfterQueue": prefs.continueAfterQueue,
            "nowPlayingVideo": prefs.nowPlayingVideo,
            "rememberMe": prefs.rememberMe,
        ]
        if let path = prefs.searchSourcePath {
            out["searchSourcePath"] = path
        }
        if let memberId, !memberId.isEmpty {
            out["activeMemberId"] = memberId
        }
        let selections = SearchHistoryStore.selections()
        if !selections.isEmpty, let encoded = encodeSelections(selections) {
            out["searchSelections"] = encoded
        }
        if let engagement = HomeTileEngagement.exportJson() {
            out["homeTileEngagement"] = engagement
        }
        if let pins = HomeSectionPinsStore.exportJson() {
            out["homeSectionPins"] = pins
        }
        if let memberId, !memberId.isEmpty {
            if let last = prefs.lastDevice {
                out["lastDevice"] = last
            }
            let pinned = PinnedDevicesStore.pinned()
            if !pinned.isEmpty {
                out["pinnedDevices"] = pinned
            }
            let offline = OfflineDownloadSync.collectForMember()
            if !offline.isEmpty, let encoded = OfflineDownloadSync.encode(offline) {
                out["offlineDownloads"] = encoded
            }
            let library = LibraryPrefsStore.load(from: prefs)
            out["libraryTab"] = LibraryPrefsStore.tabValue(for: library.filter)
            out["libraryViewMode"] = library.viewMode == .grid ? "grid" : "list"
            out["librarySortBy"] = library.sort == .recents ? "recents" : "name"
            out["librarySortOrder"] = library.sort == .recents ? "desc" : "asc"
        }
        return out
    }

    static func applyMerged(
        prefs: AppPreferences,
        merged: [String: Any],
        repository: BockMediaRepository? = nil,
        profileSwitch: Bool = false
    ) {
        if merged.isEmpty, !profileSwitch { return }
        if let v = merged["searchAllLibraries"] as? Bool {
            prefs.searchAllLibraries = v
        } else if profileSwitch {
            prefs.searchAllLibraries = true
        }
        if let path = merged["searchSourcePath"] as? String, !path.isEmpty {
            prefs.searchSourcePath = path
        } else if profileSwitch || merged.keys.contains("searchSourcePath") {
            prefs.searchSourcePath = nil
        }
        if let v = merged["downloadWifiOnly"] as? Bool {
            prefs.downloadWifiOnly = v
        } else if profileSwitch {
            prefs.downloadWifiOnly = false
        }
        if let v = merged["crossfadeSeconds"].flatMap(intPref) {
            prefs.crossfadeSeconds = min(20, max(0, v))
        } else if profileSwitch {
            prefs.crossfadeSeconds = 0
        }
        if let v = merged["continueAfterQueue"] as? String {
            prefs.continueAfterQueue = v
        } else if profileSwitch {
            prefs.continueAfterQueue = "off"
        }
        if let v = merged["nowPlayingVideo"] as? Bool {
            prefs.nowPlayingVideo = v
        } else if profileSwitch {
            prefs.nowPlayingVideo = false
        }
        if let v = merged["rememberMe"] as? Bool {
            prefs.rememberMe = v
        }
        // Do not restore activeMemberId from server prefs — profile is chosen explicitly on-device.
        if let items = decodeSelections(merged["searchSelections"]) {
            SearchHistoryStore.replaceSelections(items)
        }
        if let raw = merged["homeTileEngagement"] as? String {
            HomeTileEngagement.importJson(raw)
        }
        if let raw = merged["homeSectionPins"] as? String {
            HomeSectionPinsStore.importJson(raw)
            HomeFeedCache.invalidate()
            HomeLoadCoordinator.resetReloadWindow()
        }
        if let last = merged["lastDevice"] as? String, !last.isEmpty {
            prefs.lastDevice = last
        }
        if let pinned = decodeStringList(merged["pinnedDevices"]) {
            PinnedDevicesStore.setPinned(pinned)
        }
        if let repository {
            if let records = OfflineDownloadSync.decode(merged["offlineDownloads"]) {
                OfflineDownloadSync.applyRemote(records, repository: repository)
            } else if profileSwitch {
                OfflineDownloadSync.applyRemote([], repository: repository)
            }
            OfflineDownloadManager.shared.refresh()
        }
        if merged.keys.contains(where: { ["libraryTab", "libraryViewMode", "librarySortBy"].contains($0) }) {
            LibraryPrefsStore.applyRemote(
                tab: merged["libraryTab"] as? String,
                viewMode: merged["libraryViewMode"] as? String,
                sortBy: merged["librarySortBy"] as? String,
                sortOrder: merged["librarySortOrder"] as? String,
                prefs: prefs
            )
        }
        prefs.notifyProfilePrefsApplied()
        NotificationCenter.default.post(name: ClientPrefsSyncNotifications.prefsApplied, object: nil)
    }

    private static func encodeSelections(_ items: [SearchRecentSelection]) -> Any? {
        guard let data = try? JSONEncoder().encode(items) else { return nil }
        return try? JSONSerialization.jsonObject(with: data)
    }

    private static func decodeSelections(_ value: Any?) -> [SearchRecentSelection]? {
        guard let value else { return nil }
        guard let data = try? JSONSerialization.data(withJSONObject: value) else { return nil }
        return try? JSONDecoder().decode([SearchRecentSelection].self, from: data)
    }

    private static func decodeStringList(_ value: Any?) -> [String]? {
        guard let arr = value as? [Any] else { return nil }
        let strings = arr.compactMap { ($0 as? String)?.nilIfBlank }
        return strings.isEmpty ? nil : strings
    }

    private static func intPref(_ value: Any) -> Int? {
        if let v = value as? Int { return v }
        if let v = value as? NSNumber { return v.intValue }
        if let v = value as? Double { return Int(v.rounded()) }
        if let s = value as? String, let v = Int(s.trimmingCharacters(in: .whitespaces)) { return v }
        return nil
    }
}

private extension String {
    var nilIfBlank: String? {
        let t = trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }
}
